File: sdk/python/nexus_bus/context.py
from typing import Any

import httpx

from nexus_bus.exceptions import NexusAuthError, NexusConnectionError


class ContextClient:
    """HTTP client for the Nexus shared context store."""

    def __init__(self, client: httpx.Client, nexus_url: str) -> None:
        self._client = client
        self._api_base = f"{nexus_url.rstrip('/')}/api/v1/context"

    def get(self, correlation_id: str, key: str) -> Any | None:
        try:
            response = self._client.get(f"{self._api_base}/{correlation_id}/{key}")
        except httpx.HTTPError as exc:
            raise NexusConnectionError(
                f"Failed to read context key '{key}' for '{correlation_id}'"
            ) from exc

        if response.status_code == 404:
            return None
        if response.status_code == 401:
            raise NexusAuthError("Invalid API key for context read")
        if not response.is_success:
            raise NexusConnectionError(
                f"Context read failed with status {response.status_code}"
            )

        return response.json()

    def set(
        self,
        correlation_id: str,
        key: str,
        value: Any,
        ttl_hours: int = 24,
    ) -> dict[str, str]:
        try:
            response = self._client.put(
                f"{self._api_base}/{correlation_id}/{key}",
                json={"value": value, "ttl_hours": ttl_hours},
            )
        except httpx.HTTPError as exc:
            raise NexusConnectionError(
                f"Failed to write context key '{key}' for '{correlation_id}'"
            ) from exc

        if response.status_code == 401:
            raise NexusAuthError("Invalid API key for context write")
        if not response.is_success:
            raise NexusConnectionError(
                f"Context write failed with status {response.status_code}"
            )

        return response.json()

    def get_all(self, correlation_id: str) -> dict[str, Any]:
        try:
            response = self._client.get(f"{self._api_base}/{correlation_id}")
        except httpx.HTTPError as exc:
            raise NexusConnectionError(
                f"Failed to read context for '{correlation_id}'"
            ) from exc

        if response.status_code == 401:
            raise NexusAuthError("Invalid API key for context read")
        if not response.is_success:
            raise NexusConnectionError(
                f"Context read failed with status {response.status_code}"
            )

        data = response.json()
        return data if isinstance(data, dict) else {}

File: sdk/javascript/src/context.ts
import type { AxiosInstance } from "axios";

import {
  NexusAuthError,
  NexusConnectionError,
} from "./errors";

export class ContextClient {
  constructor(
    private readonly client: AxiosInstance,
    private readonly apiBase: string,
  ) {}

  async get<T = unknown>(
    correlationId: string,
    key: string,
  ): Promise<T | null> {
    try {
      const response = await this.client.get(
        `${this.apiBase}/${correlationId}/${key}`,
      );
      return response.data as T;
    } catch (error) {
      if (isAxiosError(error) && error.response?.status === 404) {
        return null;
      }
      if (isAxiosError(error) && error.response?.status === 401) {
        throw new NexusAuthError();
      }
      throw new NexusConnectionError(
        `Failed to read context key '${key}' for '${correlationId}'`,
      );
    }
  }

  async set(
    correlationId: string,
    key: string,
    value: unknown,
    ttlHours = 24,
  ): Promise<{ status: string; key: string }> {
    try {
      const response = await this.client.put(
        `${this.apiBase}/${correlationId}/${key}`,
        { value, ttl_hours: ttlHours },
      );
      return response.data;
    } catch (error) {
      if (isAxiosError(error) && error.response?.status === 401) {
        throw new NexusAuthError();
      }
      throw new NexusConnectionError(
        `Failed to write context key '${key}' for '${correlationId}'`,
      );
    }
  }

  async getAll(correlationId: string): Promise<Record<string, unknown>> {
    try {
      const response = await this.client.get(`${this.apiBase}/${correlationId}`);
      return response.data ?? {};
    } catch (error) {
      if (isAxiosError(error) && error.response?.status === 401) {
        throw new NexusAuthError();
      }
      throw new NexusConnectionError(
        `Failed to read context for '${correlationId}'`,
      );
    }
  }
}

function isAxiosError(
  error: unknown,
): error is { response?: { status?: number } } {
  return typeof error === "object" && error !== null && "response" in error;
}
